sdk/iot/provisioning/tests/cmocka/test_az_iot_provisioning_client.c

```diff
@@ -12,9 +12,153 @@
 
 #include <cmocka.h>
 
-int test_iot_provisioning_client()
+static const az_span test_global_device_endpoint
+    = AZ_SPAN_LITERAL_FROM_STR("global.azure-devices-provisioning.net");
+
+#define TEST_ID_SCOPE "0neFEEDC0DE"
+static const az_span test_id_scope = AZ_SPAN_LITERAL_FROM_STR(TEST_ID_SCOPE);
+
+#define TEST_REGISTRATION_ID "myRegistrationId"
+static const az_span test_registration_id = AZ_SPAN_LITERAL_FROM_STR(TEST_REGISTRATION_ID);
+
+#define TEST_USER_AGENT "c/1.2.3"
+#define TEST_REQUEST_ID "9060edd6-cc37-43d7-a96e-1bb3c3abcb82"
+
+static void test_az_iot_provisioning_client_options_default_succeed()
+{
+  az_iot_provisioning_client_options options = az_iot_provisioning_client_options_default();
+  assert_true(az_span_is_content_equal(options.user_agent, AZ_SPAN_NULL));
+}
+
+static void test_az_iot_provisioning_client_default_options_get_connect_info_succeed()
+{
+  az_iot_provisioning_client client;
+  az_result ret = az_iot_provisioning_client_init(
+      &client, test_global_device_endpoint, test_id_scope, test_registration_id, NULL);
+  assert_int_equal(AZ_OK, ret);
+
+  uint8_t client_id_buffer[128];
+  az_span client_id = AZ_SPAN_LITERAL_FROM_BUFFER(client_id_buffer);
+  az_span_set(client_id, 0xCC);
+
+  ret = az_iot_provisioning_client_id_get(&client, client_id, &client_id);
+  assert_int_equal(AZ_OK, ret);
+  assert_memory_equal(
+      az_span_ptr(test_registration_id), az_span_ptr(client_id), (size_t)az_span_length(client_id));
+  assert_int_equal(0xCC, client_id_buffer[az_span_length(client_id)]);
+
+  uint8_t username_buffer[128];
+  az_span username = AZ_SPAN_LITERAL_FROM_BUFFER(username_buffer);
+  az_span_set(username, 0xCC);
+
+  az_span expected_username
+      = AZ_SPAN_LITERAL_FROM_STR(TEST_ID_SCOPE "/registrations/" TEST_REGISTRATION_ID
+                                               "/api-version=" AZ_IOT_PROVISIONING_SERVICE_VERSION);
+
+  ret = az_iot_provisioning_client_user_name_get(&client, username, &username);
+  assert_int_equal(AZ_OK, ret);
+  assert_memory_equal(
+      az_span_ptr(expected_username), az_span_ptr(username), (size_t)az_span_length(username));
+  assert_int_equal(0xCC, username_buffer[az_span_length(username)]);
+}
+
+static void test_az_iot_provisioning_client_custom_options_get_username_succeed()
+{
+  az_iot_provisioning_client client;
+  az_iot_provisioning_client_options options = az_iot_provisioning_client_options_default();
+  options.user_agent = AZ_SPAN_FROM_STR(TEST_USER_AGENT);
+
+  az_result ret = az_iot_provisioning_client_init(
+      &client, test_global_device_endpoint, test_id_scope, test_registration_id, &options);
+  assert_int_equal(AZ_OK, ret);
+
+  uint8_t username_buffer[128];
+  az_span username = AZ_SPAN_LITERAL_FROM_BUFFER(username_buffer);
+  az_span_set(username, 0xCC);
```

Comment:
these test spans can use the new `az_span_for_test_` stuff
[Link](https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/core/core/test/inc/az_test_span.h)

Reply:
Will address in a future PR. I don't want to finalize on tests just yet since we might change the API.